Fix EZP-27528: Multiple StateGroup limitations in one policy combine with the "or" relationship instead of the "and" one

natanael89 opened the pull request:
> [EZP-27528](https://jira.ez.no/browse/EZP-27528)
#
Excerpt from the JIRA:
> When using multiple StateGroup limitations in one policy they combine using the "or" relationship instead of the "and" one. This means only one of these limitations must apply in order for the policy to work. Other limitations combine using the "and" relationship.

I found out that all the StateGroup limitations from one policy are combined together and verified against the object states. This means only one match is required for all StateGroup limitations to pass.
In this PR I changed this behaviour to at first find out which StateGroups are used for the limitation and then try to match all of them separately against the object states. If at least one of them doesn't match, then the StateGroup limitation fails.

As the ObjectStateLimitationType didn't have any tests, I added some basic tests to cover the fixed behaviour.

TODO:
- [x] reflect the new logic in `getCriterion` to make sure search queries don't try to fetch the disallowed content
- [x] _integration_ test coverage on both cases _(evaluate as used on load, and criteria as used on queries)_

natanael89 commented:
I guess the errors from Travis are not related to this PR.

bdunogier commented:
Please update the PR now that #2040 is merged, to see if it passes.

andrerom commented:
@natanael89 you need to target 6.7 branch which is the LTS for bug fixes.

natanael89 commented:
I changed the target for both of them to 6.7.

andrerom commented on eZ/Publish/Core/Limitation/ObjectStateLimitationType.php:
```diff
@@ -156,20 +158,60 @@ public function evaluate(APILimitationValue $value, APIUserReference $currentUse
+    private function areObjectStatesMatchingTheLimitation(array $objectStateIds, array $limitationValues)
+    {
+        foreach ($objectStateIds as $objectStateId) {
+            if (!in_array($objectStateId, $limitationValues, false)) {
```
Afaik you can omit the third parameter, false is the default and reading it makes me have to double check what it does :)

natanael89 replied:
Yes, it is the default, but I put it there to point out that there is no type checking on purpose - because if it is the code won't work. But I can remove it if you think that it will be more readable this way.

natanael89 commented:
I removed the unnecessary argument, added the comment and also fixed the PHPDoc so it is clear that limitation values are strings.

natanael89 commented:
Sent to QA.

natanael89 commented:
I added the TODOs suggested by @andrerom

andrerom commented on eZ/Publish/Core/Limitation/ObjectStateLimitationType.php:
```diff
@@ -236,8 +236,49 @@ public function getCriterion(APILimitationValue $value, APIUserReference $curren
+            $stateIds = array_map(function ($state) {
+                return $state->id;
+            }, $states);
+            $group = array_intersect($stateIds, $limitationValues);
```
it is a bit confusing to use the word group here ;)

natanael89 replied:
Fixed

andrerom commented on eZ/Publish/Core/Limitation/ObjectStateLimitationType.php:
```diff
@@ -236,8 +236,49 @@ public function getCriterion(APILimitationValue $value, APIUserReference $curren
+
+        if (count($groupedLimitationValues) === 1) {
+            // several limitation values: IN operation
+            return new Criterion\ObjectStateId($value->limitationValues);
```
Should this be `$groupedLimitationValues[0]` in case some of the filtering being done on that one group? Otherwise maybe comment on why we don't use the value here, if so could the count have been done on`$value->limitationValues)` ?

natanael89 replied:
It should work the same regardless if we use `$value->limitationValues` or `$groupedLimitationValues[0]`. This `if` basically means: if there is only one group found, one ObjectStateId Criterion with all the limitation values is enough (because the IN operation will be performed on them).

andrerom replied:
@natanael89 so are you saying this check could have been done before call to `groupLimitationValues ` then?

as a:
```php
    if (count($value->limitationValues) === 1) {
       // several limitation values: IN operation
       return new Criterion\ObjectStateId($value->limitationValues);
    }

```

natanael89 replied:
The check that you proposed is done in another form here: https://github.com/natanael89/ezpublish-kernel/blob/a5af4ac2731f3f5e8dee28e8a6a1bc9e2fae8b3b/eZ/Publish/Core/Limitation/ObjectStateLimitationType.php#L234. It checks if there is only one limitation value and it indeed is done before the call to `groupLimitationValues`. However, there can also be the case when there are multiple limitation values, but all of them are coming from the same state group. This case is what ` if (count($groupedLimitationValues) === 1)` is checking. It requires grouping them first though.

natanael89 replied:
Added comments to the code to help to clarify this.

andrerom commented on eZ/Publish/Core/Limitation/ObjectStateLimitationType.php:
```diff
@@ -236,8 +236,49 @@ public function getCriterion(APILimitationValue $value, APIUserReference $curren
+
+        $criterions = [];
+        foreach ($groupedLimitationValues as $limitationGroup) {
+            if (count($limitationGroup) === 1) {
```
do we need this, I know we have it other places, but logic wise both sql and solr will be fine with either case right? If so we actually don't need to optimize this here.

andrerom commented:
I still don't see why:
```php
if (count($groupedLimitationValues) === 1) {
    // one group, several limitation values: IN operation
    return new Criterion\ObjectStateId($value->limitationValues);
}
```
Is not using `return new Criterion\ObjectStateId($groupedLimitationValues[0]);`.


And I don't think we need the:
```php
if (count($limitationGroup) === 1) {
    // 1 limitation value: EQ operation
```

Afaik they cal always be IN, result will be the same.

natanael89 commented:
@andrerom It was still WIP, but now I have added all of your suggestions. Also, I have added the UnitTests for `getCriterion` and integration tests for both cases, so the PR can be reviewed again.
One thing that I'm not happy about is that here: https://github.com/ezsystems/ezpublish-kernel/pull/2035/files#diff-3560c20b3e0a970caeb79e29a1dae315R252 I'm returning a LogicalAnd which doesn't implement CriterionInterface (and the Limitation/Type interface suggests it in the PHPDoc). However, I don't see a valid alternative.
Ping @bdunogier

bdunogier commented:
> I'm returning a LogicalAnd which doesn't implement CriterionInterface (and the Limitation/Type interface suggests it in the PHPDoc).

I have checked the code, and indeed, `LogicalAnd` does not implement `CriterionInterface`. It extends the `LogicalOperator` class, that extends the `Criterion` abstract class. Other criteria, such as `ContentTypeIdentifier` extend the same `Criterion`, but *also* explicitly implement `CriterionInterface`.

I have tried adding `implements CriterionInterface` to `LogicalOperator`, and it seems okay.

@andrerom do you remember a reason that I'm forgetting ?

andrerom commented:
> @andrerom do you remember a reason that I'm forgetting ?

Well, none of them seems to implement `getSpecifications()`, so that might be the reason somehow.

andrerom commented:
@natanael89 please process with the PR despite the type hint being wrong, or change it to add `|LogicalOperator`.

If we want to fix this it seems it will need a slight bc break by introducing sub interfaces of Criterion Interface, one for logical operators _(with for instance `getCriterion()` and constants for the supported logical operators)_ and another one for attributes or whatever we call it and move _(slight BC break)_ `getSpecifications()` to that one. See the [following java classes](http://docs.oracle.com/html/E10684_08/oracle/adf/view/rich/model/Criterion.html) for examples on that.

So afaik something we can do in 7.0, unless @bdunogier thinks otherwise.

natanael89 commented:
I have added the return type. 
After this PR is merged, I will create an improvement and try to make a PR for 7.0 with your suggestion.

alongosz commented on eZ/Publish/API/Repository/Tests/Values/User/Limitation/ObjectStateLimitationTest.php:
```diff
@@ -144,4 +147,181 @@ public function testObjectStateLimitationForbid()
     }
+
+    /**
+     * Tests an ObjectStateLimitation.
```
Nitpick: I'd provide more verbose description in this case as it tests some specific forbidden condition, right?

alongosz commented on eZ/Publish/API/Repository/Tests/Values/User/Limitation/ObjectStateLimitationTest.php:
```diff
@@ -144,4 +147,181 @@ public function testObjectStateLimitationForbid()
+     * @see \eZ\Publish\API\Repository\Values\User\Limitation\ObjectStateLimitation
+     *
+     * @throws \ErrorException
+     * @expectedException \eZ\Publish\API\Repository\Exceptions\UnauthorizedException
```
Nitpick: This test method is quite long and performs multiple API calls that might throw `UnauthorizedException`, so just to be safe I'd add here `@expectedExceptionMessage` containing relevant part of expected message. Even if right now it doesn't make difference, when some other code evolves we might not notice that `UnauthorizedException` starts referring to something else. Does this make sense? :)

alongosz commented on eZ/Publish/API/Repository/Tests/Values/User/Limitation/ObjectStateLimitationTest.php:
```diff
@@ -144,4 +147,181 @@ public function testObjectStateLimitationForbid()
+
+        $removePolicy = null;
+        foreach ($role->getPolicies() as $policy) {
+            if ('content' != $policy->module || 'remove' != $policy->function) {
```
Nitpick: please use `!==` operator.

alongosz commented on eZ/Publish/API/Repository/Tests/Values/User/Limitation/ObjectStateLimitationTest.php:
```diff
@@ -144,4 +147,181 @@ public function testObjectStateLimitationForbid()
+
+        if (null === $removePolicy) {
+            throw new \ErrorException('No content:remove policy found.');
+        }
```
Please replace with `assertNotNull` method call.

alongosz commented on eZ/Publish/API/Repository/Tests/Values/User/Limitation/ObjectStateLimitationTest.php:
```diff
@@ -144,4 +147,181 @@ public function testObjectStateLimitationForbid()
+                        $lockedState,
+                        $defaultStateFromAnotherGroup,
+                    ),
+                )
```
Nitpick: please use short array syntax as this is new code.

alongosz commented on eZ/Publish/Core/Limitation/Tests/ObjectStateLimitationTypeTest.php:
```diff
@@ -0,0 +1,310 @@
+        parent::setUp();
+
+        $this->objectStateHandlerMock = $this->getMock(
+            'eZ\\Publish\\SPI\\Persistence\\Content\\ObjectState\\Handler',
```
Nitpick: Please use `Handler::class`

alongosz commented on eZ/Publish/Core/Limitation/Tests/ObjectStateLimitationTypeTest.php:
```diff
@@ -0,0 +1,310 @@
+        $this->objectStateHandlerMock = $this->getMock(
+            'eZ\\Publish\\SPI\\Persistence\\Content\\ObjectState\\Handler',
+            array(),
+            array(),
```
Nitpick: please use short array syntax as it's a new code.

alongosz commented on eZ/Publish/Core/Limitation/Tests/ObjectStateLimitationTypeTest.php:
```diff
@@ -0,0 +1,310 @@
+            $this->getUserMock()
+        );
+
+        self::assertInstanceOf('\eZ\Publish\API\Repository\Values\Content\Query\Criterion\ObjectStateId', $criterion);
```
Nitpick: please use `ObjectStateId::class`

alongosz commented on eZ/Publish/Core/Limitation/Tests/ObjectStateLimitationTypeTest.php:
```diff
@@ -0,0 +1,310 @@
+            $this->getUserMock()
+        );
+
+        self::assertInstanceOf('\eZ\Publish\API\Repository\Values\Content\Query\Criterion\ObjectStateId', $criterion);
```
Nitpick: please use `ObjectStateId::class`

alongosz commented on eZ/Publish/Core/Limitation/Tests/ObjectStateLimitationTypeTest.php:
```diff
@@ -0,0 +1,310 @@
+            $this->getUserMock()
+        );
+
+        self::assertInstanceOf('\eZ\Publish\API\Repository\Values\Content\Query\Criterion\LogicalAnd', $criterion);
```
Nitpick: please use `LogicalAnd::class`

natanael89 commented:
@alongosz thanks for your suggestion. I have added them to the code.
I'm sending it to the QA.
